File: NodeJS-Starter-V1/apps/backend/src/api/routes/customers.py
"""Customers API routes."""
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from src.config.database import get_db
from src.db.demo_models import Customer as CustomerModel
from src.db.schemas import Customer, CustomerCreate, CustomerUpdate, PaginatedResponse

router = APIRouter(prefix="/api/customers", tags=["customers"])


@router.get("", response_model=PaginatedResponse)
async def list_customers(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    search: str | None = None,
    is_active: bool | None = None,
    db: AsyncSession = Depends(get_db),
):
    """List customers with pagination and filters."""
    # Build query
    query = select(CustomerModel)

    # Apply filters
    if search:
        search_filter = f"%{search}%"
        query = query.where(
            (CustomerModel.company_name.ilike(search_filter)) |
            (CustomerModel.customer_number.ilike(search_filter)) |
            (CustomerModel.contact_name.ilike(search_filter)) |
            (CustomerModel.email.ilike(search_filter))
        )

    if is_active is not None:
        query = query.where(CustomerModel.is_active == is_active)

    # Get total count
    count_query = select(func.count()).select_from(query.subquery())
    result = await db.execute(count_query)
    total = result.scalar_one()

    # Apply pagination
    query = query.offset((page - 1) * page_size).limit(page_size)
    query = query.order_by(CustomerModel.created_at.desc())

    # Execute query
    result = await db.execute(query)
    customers = result.scalars().all()

    return {
        "items": [Customer.model_validate(c) for c in customers],
        "total": total,
        "page": page,
        "page_size": page_size,
        "total_pages": (total + page_size - 1) // page_size,
    }


@router.get("/{customer_id}", response_model=Customer)
async def get_customer(
    customer_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get a single customer by ID."""
    query = select(CustomerModel).where(CustomerModel.id == customer_id)
    result = await db.execute(query)
    customer = result.scalar_one_or_none()

    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")

    return Customer.model_validate(customer)


@router.post("", response_model=Customer, status_code=201)
async def create_customer(
    customer_data: CustomerCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new customer."""
    # Prepare customer data
    data_dict = customer_data.model_dump(exclude_unset=True)

    # Handle name/company_name aliasing
    if "name" in data_dict and not data_dict.get("company_name"):
        data_dict["company_name"] = data_dict.pop("name")
    elif "name" in data_dict:
        data_dict.pop("name")  # Remove alias if company_name exists

    # Handle postal_code/postcode aliasing
    if "postal_code" in data_dict and not data_dict.get("postcode"):
        data_dict["postcode"] = data_dict.pop("postal_code")
    elif "postal_code" in data_dict:
        data_dict.pop("postal_code")

    # Auto-generate customer number if not provided
    if not data_dict.get("customer_number"):
        # Get latest customer number
        query = select(CustomerModel.customer_number).order_by(
            CustomerModel.created_at.desc()
        ).limit(1)
        result = await db.execute(query)
        latest_number = result.scalar_one_or_none()

        if latest_number and latest_number.startswith("CUST-"):
            try:
                last_num = int(latest_number.split("-")[1])
                data_dict["customer_number"] = f"CUST-{last_num + 1:06d}"
            except (ValueError, IndexError):
                data_dict["customer_number"] = "CUST-000001"
        else:
            data_dict["customer_number"] = "CUST-000001"

    # Validate required fields
    if not data_dict.get("company_name"):
        raise HTTPException(status_code=400, detail="company_name or name is required")

    # Check if customer number already exists
    existing_query = select(CustomerModel).where(
        CustomerModel.customer_number == data_dict["customer_number"]
    )
    result = await db.execute(existing_query)
    if result.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Customer number already exists")

    # Check if email already exists (if email is provided)
    if data_dict.get("email"):
        email_query = select(CustomerModel).where(
            CustomerModel.email == data_dict["email"]
        )
        result = await db.execute(email_query)
        if result.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Email already exists")

    # Remove fields not in the model
    model_fields = {"customer_number", "company_name", "contact_name", "email",
                   "phone", "address", "city", "state", "postcode", "is_active",
                   "organization_id"}
    filtered_data = {k: v for k, v in data_dict.items() if k in model_fields}

    # Create customer
    customer = CustomerModel(**filtered_data)
    db.add(customer)
    await db.commit()
    await db.refresh(customer)

    return Customer.model_validate(customer)


@router.put("/{customer_id}", response_model=Customer)
async def update_customer(
    customer_id: UUID,
    customer_data: CustomerUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update a customer."""
    # Get existing customer
    query = select(CustomerModel).where(CustomerModel.id == customer_id)
    result = await db.execute(query)
    customer = result.scalar_one_or_none()

    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")

    # Update fields
    update_data = customer_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(customer, field, value)

    await db.commit()
    await db.refresh(customer)

    return Customer.model_validate(customer)


@router.delete("/{customer_id}", status_code=204)
async def delete_customer(
    customer_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Soft delete a customer (set is_active to False)."""
    # Get existing customer
    query = select(CustomerModel).where(CustomerModel.id == customer_id)
    result = await db.execute(query)
    customer = result.scalar_one_or_none()

    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")

    # Soft delete
    customer.is_active = False
    await db.commit()

    return None

File: apps/web/lib/api/customers.ts
import { apiClient } from "./client";

/**
 * Customer interface
 */
export interface Customer {
  id: string;
  customer_number: string;
  company_name: string;
  contact_name: string;
  email: string;
  phone?: string;
  address?: string;
  city?: string;
  state?: string;
  postcode?: string;
  country?: string;
  xero_contact_id?: string;
  xero_synced_at?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * Customer create request
 */
export interface CustomerCreate {
  company_name: string;
  contact_name: string;
  email: string;
  phone?: string;
  address?: string;
  city?: string;
  state?: string;
  postcode?: string;
  country?: string;
}

/**
 * Customer update request
 */
export interface CustomerUpdate {
  company_name?: string;
  contact_name?: string;
  email?: string;
  phone?: string;
  address?: string;
  city?: string;
  state?: string;
  postcode?: string;
  country?: string;
  is_active?: boolean;
}

/**
 * Customer list params
 */
export interface CustomerListParams {
  page?: number;
  page_size?: number;
  search?: string;
  is_active?: boolean;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

/**
 * Paginated customer response
 */
export interface PaginatedCustomers {
  items: Customer[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/**
 * Customers API client
 */
export const customersApi = {
  /**
   * List customers with pagination, search, and filters
   */
  async list(params: CustomerListParams = {}): Promise<PaginatedCustomers> {
    const queryParams = new URLSearchParams();
    if (params.page) queryParams.append("page", params.page.toString());
    if (params.page_size) queryParams.append("page_size", params.page_size.toString());
    if (params.search) queryParams.append("search", params.search);
    if (params.is_active !== undefined) queryParams.append("is_active", params.is_active.toString());
    if (params.sort_by) queryParams.append("sort_by", params.sort_by);
    if (params.sort_order) queryParams.append("sort_order", params.sort_order);

    return apiClient.get<PaginatedCustomers>(`/api/customers?${queryParams.toString()}`);
  },

  /**
   * Get a single customer by ID
   */
  async get(id: string): Promise<Customer> {
    return apiClient.get<Customer>(`/api/customers/${id}`);
  },

  /**
   * Create a new customer
   */
  async create(data: CustomerCreate): Promise<Customer> {
    return apiClient.post<Customer>("/api/customers", data);
  },

  /**
   * Update an existing customer
   */
  async update(id: string, data: CustomerUpdate): Promise<Customer> {
    return apiClient.put<Customer>(`/api/customers/${id}`, data);
  },

  /**
   * Soft delete a customer (set is_active = false)
   */
  async delete(id: string): Promise<void> {
    return apiClient.delete(`/api/customers/${id}`);
  },
};
